### PR TITLE
Project 2

### PR DESCRIPTION
**Pull request by mohamed20163858**
Build `CapitalizeDecorator` and `TrimmerDecorator` to capitalize or trim the name attribute inside the `Person` class without needing to extend the class methods that much using the Decorator design pattern principle. this is done by creating the `Nameable` class as a parent class for `Person` and all defined Decorators then creating a base `Decorator` class that is directly the parent of the `CapitalizeDecorator` and `TrimmerDecorator` classes and then we can use them to pass any person object to capitalize or trim the name attribute inside the person. and here is a screenshot of the results of running this script:- 
``` 
require './person'
require './capitalize_decorator'
require './trimmer_decorator'
person = Person.new(22, 'maximilianus')
person.correct_name
capitalized_person = CapitalizeDecorator.new(person)
capitalized_person.correct_name
capitalized_trimmed_person = TrimmerDecorator.new(capitalized_person)
capitalized_trimmed_person.correct_name
``` 
image

### CODE REVIEW

File: decorator.rb [hunk 1]
```diff
@@ -0,0 +1,11 @@
+require './nameable'
+class Decorator < Nameable
```
Comment (zieeco) [inline decorator.rb:2]: - Please, let's also add the ` attr_accessor :nameable` here in order to access it.

File: trimmer_decorator.rb [hunk 2]
```diff
@@ -0,0 +1,7 @@
+require './decorator'
+class TrimmerDecorator < Decorator
+  def correct_name
+    str = @nameable.correct_name
+    str.length > 10 ? str[0, 10] : str
```
Comment (zieeco) [inline trimmer_decorator.rb:5]: - Hey, the name should return a maximum of 10 characters, ie [0, 9] as in your case is returning 11 characters. please let's update it.

**Review by zieeco — CHANGES_REQUESTED** [review]
Hi @mohamed20163858 ,


Great job so far! 👏 
You've done well, however, there are some issues that you still need to work on to go to the next project, but you are almost there!


## To highlight 🎯

- No linter errors.✔️
- Nice class decoration. ✔️
- Descriptive PR. ✔️


## Required Changes ♻️
_Check the comments under the review._





## Optional suggestions
- N/A

**_Every comment with the **[OPTIONAL]** prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._**


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: trimmer_decorator.rb (hunk 2 above)
Comment (mohamed20163858) [reply trimmer_decorator.rb:5]: @zieeco actually with all respect you didn't notice a comma there is a difference between `str[0..9]` and `str[0,9]` the  first syntax means get all chars that begin at index 0  till index  9 so it has a total of 10 chars, but `str[0,9]` means fetch the substring that begins at index 0 and has a total length = 9  that is it

Comment (mohamed20163858) [reply trimmer_decorator.rb:5]: @zieeco  an example you can find in that [article](https://www.rubyguides.com/2018/01/ruby-string-methods/#How_to_Extract_a_Substring) 
image

File: decorator.rb (hunk 1 above)
Comment (mohamed20163858) [reply decorator.rb:2]: @zieeco ok got it but I think it wasn't stated in project requirements

**Review by mikethreels — APPROVED** [review]
Hi Mohamed, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights
- linters are passing
## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**